R/time_binning.R

```diff
@@ -0,0 +1,325 @@
+#' Assign fossil occurrences to time bins
+#'
+#' A function to assign fossil occurrences to specified time bins based on
+#' different approaches commonly applied in palaeobiology.
+#'
+#' @param occdf \code{dataframe}. A dataframe of the fossil occurrences you wish
+#' to bin. This dataframe should contain  the following named columns:
+#' "max_ma" and "min_ma". These columns may be either `numeric` or `character`
+#' values. Ages given in `numeric` form are preferred for bin assignment.
+#' However, if interval names are provided, the function will try to match names
+#' to the Geological Timescale 2012/2020 (depending on user specification) to
+#' generate `numeric` ages for occurrences (see "return_error" for additional information).
+#' @param bins \code{dataframe}. A dataframe of the bins that you wish to allocate
+#' fossil occurrences to. This dataframe must contain at least the following named
+#'  columns: "bin", "max_ma" and "min_ma", and must be `numeric` values.
+#' @param method \code{character}. The method desired for binning fossil occurrences.
+#' Currently, five methods exist in this function: "mid", "majority", "all", "random",
+#' and "point". See Details for a description of each.
+#' @param reps \code{numeric}. A non-negative `numeric` specifying the number of
+#' replications for sampling. This argument is only useful in the case of the "random"
+#' or "point" method being specified in the `method` argument. Defaults to 100.
+#' @param scale \code{character}. Specify the desired geological timescale to be used
+#' "GTS2020" or "GTS2012". This argument is only relevant if "min_ma" and "max_ma"
+#' columns are `character` values of interval names. Available interval names can be
+#' accessed via the call GTS2020$interval_name or GTS2012$interval_name. "GTS2020" is
+#' the default option.
+#' @param return_error \code{logical}. Should a vector of numbers be returned to flag
+#' the rows of the `occdf` that cannot be matched to `character` interval names? This is
+#' only relevant if `occdf$max_ma` and `occdf$min_ma` are `character` values.
+#'
+#' @return A \code{dataframe} of the original input `occdf` with the following appended
+#' columns: occurrence id (`id`), number of bins the occurrence age range covers (`n_bins`), bin assignment
+#' (`bin_assignment`), and bin midpoint (`bin_midpoint`). In the case of the "random" method,
+#' a \code{list} is appended for the bin assignment and midpoint columns. If `method` is
+#' specified as "point", only point age estimates are appended (`point_estimates`) as a
+#' \code{list} in the dataframe.
```

Comment:
Hi Will, I think Chris and I actually agree with you that we would prefer this as lists of dataframes. However, we are concerned about how heavy this might be. Imagine if someone was using the whole PBDB and 1000 reps.... this would be very heavy. The current implementation allows the data to be conveniently accessed in a way that minimises consumed memory:

Reply:
Yeah, I totally understand the memory implications, but I'm really convinced that current format will be much harder for users to use efficiently/effectively. For most analyses I'm familiar with, you want the age/bin data in a single column within the dataframe. If we want to keep the current format, we at the very least should have some code in the example (or a vignette) to show how users would convert from the current format to an easier format.

Also, would it be possible to figure out what the memory difference would be for the worst case scenario (the entire PBDB)? I've used large lists of dataframes before, and they don't end up being as large as one might expect. And if someone is already using the entire PBDB, maybe they would be ok with large memory usage?

Reply:
I think we went over this in the meeting but to confirm here, I've now updated this. The function returns lists of dataframes for each replication now.